### PR TITLE
CMR-7080 Creating a new access-control route to search for s3 buckets for providers

### PR DESCRIPTION
**Pull request by jaybarra**

### CODE REVIEW

File: access-control-app/src/cmr/access_control/services/acl_service.clj [hunk 1]
```diff
@@ -357,3 +359,49 @@
...
+                    provider-ids)]
+    (filter #(some #{:update}
+                   (first (vals (get-permissions
+                                 context
+                                 {:user_id user-id
+                                  :target "INGEST_MANAGEMENT_ACL"
+                                  :provider %}))))
+            providers)))
+
+(defn- fetch-s3-buckets-for-providers
```
Comment (jaybarra) [inline access-control-app/src/cmr/access_control/services/acl_service.clj:379]: replace query by provider to query by groups

**Review by daniel-zamora — APPROVED** [review]

File: access-control-app/docs/api.md [hunk 2]
```diff
@@ -873,6 +875,41 @@ Content-Type: application/json;charset=ISO-8859-1
 ** `user_id` - The user whose permissions will be computed.
 ** `user_type` - Either "guest" or "registered".
 
+### <a name="get-s3-buckets"></a> User Access to S3 Buckets
+
+This endpoint will return a list of S3 buckets a user has access to. If a list of providers is included in the request, the list of S3 buckets will be filtered to only include S3 buckets contained by collections within those providers.
```
Comment (ygliuvt) [inline access-control-app/docs/api.md:880]: Add some language indicating that the result format in JSON.

File: access-control-app/docs/api.md [hunk 3]
```diff
@@ -873,6 +875,41 @@ Content-Type: application/json;charset=ISO-8859-1
...
+* Optional
+** `provider[]` - A list of provider ids to filter the S3 bucket and object prefix names by. If not specified, all providers will be included.
+
+
+Example request:
+```
+curl -i -H "Echo-Token: XXXX" "%CMR-ENDPOINT%/s3-buckets?user_id=user1
+
+HTTP/1.1 200 OK
+Content-Length: 81
```
Comment (ygliuvt) [inline access-control-app/docs/api.md:896]: Include Content-Type here to make it clear that the result format is application/json.

File: access-control-app/src/cmr/access_control/services/acl_service.clj [hunk 4]
```diff
@@ -357,3 +360,39 @@
...
+         flatten
+         set
+         sort)))
+
+(defn s3-buckets-for-user
+  "Returns a list of s3 buckets and object prefix names by provider."
+  [context user provider-ids]
+  (let [groups (auth-util/get-sids context user)
+        providers (if (empty? provider-ids)
+                    (map :provider-id (mdb/get-providers context))
```
Comment (ygliuvt) [inline access-control-app/src/cmr/access_control/services/acl_service.clj:394]: providers should be cached in access-control. We can't afford to call metadata-db to retrieve providers on every get s3 prefix call that does not specify a provider.

File: access-control-app/src/cmr/access_control/services/acl_service.clj [hunk 5]
```diff
@@ -357,3 +360,39 @@
     (if (or (nil? user-token)(re-matches #"^[\s\t]*$" user-token))
       (auth-util/get-sids context :guest)
       (auth-util/get-sids context (tokens/get-user-id context user-token)))))
+
+(defn- fetch-s3-buckets-by-groups
+  [context groups providers]
+  (let [base-query (gc/or-conds (map #(qm/string-condition :permitted-group-ids
```
Comment (ygliuvt) [inline access-control-app/src/cmr/access_control/services/acl_service.clj:366]: why do we match group-id with wildcards?

File: access-control-app/src/cmr/access_control/services/parameter_validation.clj [hunk 6]
```diff
@@ -108,3 +108,10 @@
   "Throws service errors if any invalid params or values are found."
   [params]
   (validate-params params :user-token))
+
+(defn validate-s3-buckets-params
+  "Throws service errors if any invalid params or values are found."
+  [params]
+  (validate-params params :user_id :provider)
+  (when-not (contains? params :user_id)
```
Comment (ygliuvt) [inline access-control-app/src/cmr/access_control/services/parameter_validation.clj:116]: change this to something like `(when (clojure.string/blank? (:user_id params))` to be more robust.

File: system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        :provider ["PROV2"]}
+       ["s3" "s3://aws.example-2.com"]
+
+       "user filtering on mixed access to providers"
+       {:user_id "user2"
+        :provider ["PROV2" "PROV3"]}
+       ["s3" "s3://aws.example-2.com"]))
+
+    (testing "validation"
+      (testing "missing user param"
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj:177]: Add a test for non-existent user.

File: system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+   [cmr.transmit.config :as transmit-config]))
+
+(use-fixtures :each (ingest/reset-fixture {"provguid1" "PROV1"
+                                           "provguid2" "PROV2"
+                                           "provguid3" "PROV3"}
+                                          {:grant-all-search? false
+                                           :grant-all-ingest? false}))
+
+(deftest s3-buckets-route-test
+  (let [all-prov-update-group-id (echo-util/get-or-create-group
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj:25]: Add comments on what groups is for what purpose and why you grant certain permissions. I find it hard to follow the permissions being granted. This is only related to searching aspect of the ACL permissions, why do we manipulating the ingest ACL permissions in the test as well?

File: system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+       ;; Admin user cases
+       "user with access to all providers"
+       {:user_id "user1"}
+       ["s3"
+        "s3://aws.example-1.com"
+        "s3://aws.example-2.com"
+        "s3://aws.example-3.com"]
+
+       "non-existant provider"
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/access_control/s3_bucket_route_test.clj:137]: Fix typos of `existent` throughout the test.

**Review by ygliuvt — CHANGES_REQUESTED** [review]
Need to add caching of providers, address some questions.

File: access-control-app/src/cmr/access_control/services/acl_service.clj (hunk 5 above)
Comment (jaybarra) [reply access-control-app/src/cmr/access_control/services/acl_service.clj:366]: This was copied from another query that did something similar, it's probably not necessary.

**Review by ygliuvt — APPROVED** [review]
